extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessorFactory.java

```diff
@@ -56,26 +56,26 @@
 @JsonTypeName("window")
 public class WindowOperatorQueryFrameProcessorFactory extends BaseFrameProcessorFactory
 {
-  private final WindowOperatorQuery query;
+  private final QueryContext queryContext;
   private final List<OperatorFactory> operatorList;
   private final RowSignature stageRowSignature;
 
   @JsonCreator
   public WindowOperatorQueryFrameProcessorFactory(
-      @JsonProperty("query") WindowOperatorQuery query,
+      @JsonProperty("queryContext") QueryContext queryContext,
       @JsonProperty("operatorList") List<OperatorFactory> operatorFactoryList,
       @JsonProperty("stageRowSignature") RowSignature stageRowSignature
   )
   {
-    this.query = Preconditions.checkNotNull(query, "query");
+    this.queryContext = Preconditions.checkNotNull(queryContext, "query");
     this.operatorList = Preconditions.checkNotNull(operatorFactoryList, "bad operator");
     this.stageRowSignature = Preconditions.checkNotNull(stageRowSignature, "stageSignature");
   }
 
-  @JsonProperty("query")
-  public WindowOperatorQuery getQuery()
+  @JsonProperty("queryContext")
```

Comment:
Would this not break existing queries which are running on with an older controller/worker?

Reply:
Can you elaborate the scenario? Is it when an upgrade happens during the worker jvm is starting up?

Reply:
If a JSON containing the key query instead of queryContext is sent by an older controller, it would not be understood by a worker with this change. The worker would not be able to seralize the FrameProcessorFactory.

Reply:
When can a controller version be different from the worker version? When the middle manager is upgraded during the worker jvm startup?

Reply:
Also, what would be the way to handle it? I guess passing both params and then removing the unused one in a later release also wouldn't work.

Reply:
This would cause MSQ queries with window functions to fail, I'm not sure if we are okay with breaking changes to window functions or its fine since it is still experimental.

Reply:
> This would cause MSQ queries with window functions to fail, I'm not sure if we are okay with breaking changes to window functions or its fine since it is still experimental.

Window functions are going to be GA as part of Druid 31 release.

> I don't think there is a backward compatible way to remove query

Yeah..

How do we generally make any changes to any other processor factory classes while ensuring backward compatibility? It doesn't seem like a trivial thing to achieve while ensuring clean code.

Is this something we can put in the release notes section? I think it's better to break backward compatibility sooner than later, if it's inevitable. Thoughts? cc: @kgyrtkirk @cryptoe

Reply:
Window functions are GA in druid 31. We should not break backward compatibility in the next release. 
The way to do it now is to 
1. Mark the older queryContext deprecated. Add a release note in the developer section basically mentioning that we will remove this parameter in druid 33. 
2. The current code handles both query/query context and takes certain assumptions. Preference given to query
3. We remove the deprecated param and the scafolding code in druid 33.

Reply:
regardless of that I think all changes outside WindowOperatorQueryFrameProcessorFactory could still be done
the orig [comment](https://github.com/apache/druid/pull/17158#discussion_r1793049861) suggested to flatten the needed config into a config object and not pass even the context map - that's also too generic.

Reply:
@kgyrtkirk Have re-opened the PR and have reverted the changes for WindowOperatorQueryFrameProcessorFactory's constructor params. Appreciate your review, thanks!